src/core/events/events.rs

```diff
@@ -0,0 +1,173 @@
+/// which the event sections originate. The second one is the field name of a
+/// given (collector) event field.
+pub(crate) struct Event(HashMap<String, HashMap<String, EventField>>);
+
+impl Event {
+    pub(crate) fn new() -> Event {
+        Event(HashMap::new())
+    }
+
+    /// Get the event len, aka. its # of fields.
+    pub(crate) fn len(&self) -> usize {
+        let mut len = 0;
+        for fields in self.0.values() {
+            len += fields.len();
+        }
+        len
+    }
+
+    /// Get a reference to an event field by its owner and key.
+    pub(crate) fn get<T: 'static>(&self, owner: &str, key: &str) -> Option<&T> {
+        if let Some(section) = self.0.get(&owner.to_string()) {
+            if let Some(field) = section.get(&key.to_string()) {
+                return field.val.as_any().downcast_ref::<T>();
+            }
+        }
+        None
+    }
+
+    /// Insert a new event field into an event.
+    pub(crate) fn insert(&mut self, key: &str, val: EventField) {
+        let key = key.to_string();
+
+        if !self.0.contains_key(&key) {
+            self.0.insert(key.clone(), HashMap::new());
+        }
+
+        // Unwrap can't fail as we checked the key exists in the above block.
+        let map = self.0.get_mut(&key).unwrap();
+        map.insert(val.key.clone(), val);
+    }
+
+    pub(crate) fn to_json(&self) -> serde_json::Value {
+        self.into()
+    }
+}
+
+// This allows converting an Event to a serde_json::Value in the
+// Event::to_json() helper.
+impl From<&Event> for serde_json::Value {
+    fn from(f: &Event) -> Self {
+        let mut event = serde_json::Map::new();
+
+        for (key, owner) in f.0.iter() {
+            let mut section = serde_json::Map::new();
+
+            for (key, field) in owner.iter() {
+                section.insert(key.clone(), field.val.to_json());
+            }
+
+            event.insert(key.clone(), serde_json::Value::Object(section));
+        }
+
+        serde_json::Value::Object(event)
+    }
+}
+
+/// Event fields are the events building blocks. They hold per-type data.
+pub(crate) struct EventField {
+    key: String,
+    val: Box<dyn EventFieldType>,
+}
+
+impl EventField {
+    pub(crate) fn new(key: &str, val: Box<dyn EventFieldType>) -> EventField {
+        EventField {
+            key: key.to_string(),
+            val,
+        }
+    }
+}
+
+/// Wrapper to easily create a new event field to insert into an event.
+///
+/// `event_field!("key", 42);`
+#[macro_export]
+macro_rules! event_field {
+    ($key:expr, $val:expr) => {
+        EventField::new($key, Box::new($val))
+    };
+}
+
+/// Implementation of an event field type, used to hold the actual data and
+/// provide helpers to serialize/deserialize it.
+pub(crate) trait EventFieldType {
+    fn name(&self) -> &'static str;
+    fn as_any(&self) -> &dyn Any;
+    fn to_json(&self) -> serde_json::Value;
+    fn from_json(from: serde_json::Value) -> Result<Self>
```

Comment:
I'm thinking about how the `from_json` side will work.
Although we have `from_json` in the `EventFieldType` trait, we will need to call it on the specific type. So we need a way to get the type from the owner and field_name.
A possible way would be to convert the `EventUnmarshaler` into a trait, something like:
```rust
trait EventUnmarshaler {
   type Field: EventFieldType;
   fn from_raw(&Vec[u8]) -> Result<Field>;
   fn from_json(serde_json::Value]) -> Result<Field>;
}
```
WDYT?

Reply:
Sorry I think I wrote some nonsense which made it more confusing.
I didn't mean to tie everything at the `EventFieldType` level but at the section level.

> I haven't found yet a perfect representation for the json to event translation. The issue we have is there is no serde_json::Value to EventFieldType relationship (same with raw sections). 

There is: `EventFieldType::from_json(from: serde_json::Value) -> Result<Self>` . Am I missing something?


>  What we have is raw section 1:1 internal collector section representation 1:n section n:1 event. 

I don't understand what's the differnt between "section" and "internal collector section representation".


What I see is that: for each `BpfEventOwner` we have an `Unmarshaler` capable of generating a `Vec[EventField]` which is then converted into an internal representation of an event section such as:
```rust
HashMap::from([
    ("owner", HashMap::from([
        ("field0_name", field0_val),
        ("field1_name", field1_val)
    ]),
])
``` 
Which is trivially converted into json:
```
"owner": {
    "field0_name": field0_val,
    "field1_name": field1_val,
}
```
Mainly because fieldN_val implement `to_json()`

To keep things simple, I would just mimic this design on the json side and have:
```rust
trait EventUnmarshaler {
   fn from_raw_section(&BpfRawSection, &mut Vec<EventField>) -> Result<()>
   fn from_json_section(&json::Map, &mut Vec<EventField>) -> Result<()>
}
```
Following the example of `skb_tracing` we would have something like:

```rust
impl Collector for SkbTrackingCollector {  
    fn init(...events: &mut BpfEvents) {
        events.register_unmarshaller(SkbTrackingUnmarshaler{})
    }  
}

struct SkbTrackingUnmarshaler{}:

impl EventUnmarshaler  for SkbTrackingUnmarshaler {
fn from_raw_section(section: &BpfRawSection, fields: &mut Vec<EventField>) -> Result<()> {
    // some integrity checks
    // use an internal struct that happens to have the same binary layout as the RawSection (nice implementation detail but not important from an API PoV)
    let field0_val = [...]; // Implements EventFieldType
    let field1_val = [...]; // Implements EventFieldType
    fields.push(event_field!("field0_name", field0_val));
    fields.push(event_field!("field1_name", field1_val));
}

fn from_json_section(map: &json::Map, fields: &mut Vec<EventField>) -> Result<()> {
    // some integrity checks
    // Maybe perform some manual complex parsing.
    let field0_val = manually_parse_field0(map.get("field0_name"));  // Implements EventFieldType
    let field1_val = manually_parse_field1(map.get("field1_name"));  // Implements EventFieldType
    
    // Maybe create an intermediate object that uses serde::Deserialize to perform some semi-automatic deserialization. Maybe it's even the same object that we use for RawSection deserialization, who knows.
    // But ultimately...
    fields.push(event_field!("field0_name", field0_val));
    fields.push(event_field!("field1_name", field1_val));
} 
}
```

Each module defines how it's section is unmarshaled from `BpfRawSection -> Vec<EventField> ` and from `json::Map -> Vec<EventField>`

This would allow collectors to implement arbitrarily complex json_to_vector_of_EventFields unmarshalers.

One step further would be to implement serde:Deserialize for our binary TLV format so a single struct can, indeed, be used to deserialize both from binary and from json.

Reply:
BTW, I'm not arguing for doing all this in this series but I do think we should at least agree on how we are going to do it so that we minimize the API changes in the future.

Essentially the problem is that currently the serialization is done at the `EventFieldType` level but deserialization cannot be done at that level unless we map names to types somehow. Having serialization and deserialization at different levels is a bit ugly IMO but doable.

A possible way to move serialization one level "up", i.e: from `EventField` to `Event` (which is a small variation of the above snippet) would be to have something like:



```rust
trait Event {
   fn from_raw_section(&BpfRawSection) -> Result<Self>
   fn from_json(&serde::Value) -> Result<Self>
   fn to_json(&self) -> Result<serde::Value>
   fn to_fields(&self, mut &Vec[EventFields]) -> Result<()>
}

// in mycollector.rs
// We could use the same object for json and binary or not.
#[derive(Default, Deserialize, Serialize)]
#[repr(C, packed)]
struct MyEventData {
    data1: u8,
    data2: u32,
}

// Not sure if we need the wrapper or not
struct MyEvent {
    data: MyEventData,
}

impl Event for MyEvent {
   // Pretty much what we have now...
   fn from_raw_section(raw_section: &BpfRawSection) -> Result<Self> {
       let mut data = MyEventData::default();
       plain::copy_from_bytes(&mut data, &raw_section.data);
       Ok(MyEvent{data})
   }
   
    // ... but split in two parts
   fn to_fields(&self, fields mut &Vec[EventFields]) -> Result<()> {
        fields.push(event_field!("data1", self.data.data1));
        fields.push(event_field!("data2", self.data.data1));
   }
   
   // to_json is optional!! The default implementation calls fields() and perform field-level marshalling just as we do now into a json map. That way we don't _need_ to implement Serialize.
   // fn to_json(&self) -> Result<serde::Value>{...} 
   
    // However, from_json needs to be implemented. There is no way we can write a default implementation becasuse we don't know the types.  
   fn from_json(json_data: &serde::Value) -> Result<Self> {
       let mut data = MyEventData::default();
       serde_json::from_json(&mut data, json_data);
       Ok(MyEvent{data})
   }
}

```

Reply:
Wondering if @vlrpl has any thoughts.